File: backend/app/models/audit_log.py
"""Audit log ORM model."""
from __future__ import annotations

import enum
from datetime import datetime
from typing import Optional, TYPE_CHECKING

from sqlalchemy import DateTime, Enum as SAEnum, ForeignKey, Integer, String, Text, func
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.base import Base, IdMixin

if TYPE_CHECKING:
    from app.models.user import User


class AuditAction(str, enum.Enum):
    LOGIN = "login"
    CREATE_TASK = "create_task"
    UPDATE_TASK = "update_task"
    DELETE_TASK = "delete_task"
    SUBMIT_TASK = "submit_task"
    APPROVE_TASK = "approve_task"
    REJECT_TASK = "reject_task"
    COMPLETE_TASK = "complete_task"


class AuditLog(Base, IdMixin):
    __tablename__ = "audit_logs"

    user_id: Mapped[Optional[int]] = mapped_column(
        ForeignKey("users.id", ondelete="SET NULL"),
        nullable=True,
        index=True,
    )
    action: Mapped[AuditAction] = mapped_column(
        SAEnum(AuditAction, name="audit_action", values_callable=lambda x: [e.value for e in x]),
        nullable=False,
        index=True,
    )
    entity_type: Mapped[str] = mapped_column(String(50), nullable=False, index=True)
    entity_id: Mapped[Optional[int]] = mapped_column(Integer, nullable=True, index=True)
    details: Mapped[Optional[str]] = mapped_column(Text, nullable=True)

    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        nullable=False,
        index=True,
    )

    user: Mapped[Optional["User"]] = relationship("User", back_populates="audit_logs")

File: frontend/src/types/auditLog.ts
import type { UserSummary } from './auth';

export const AUDIT_ACTIONS = [
  'login',
  'create_task',
  'update_task',
  'delete_task',
  'submit_task',
  'approve_task',
  'reject_task',
  'complete_task',
] as const;

export type AuditAction = (typeof AUDIT_ACTIONS)[number];

export interface AuditLog {
  id: number;
  user_id: number | null;
  user: UserSummary | null;
  action: AuditAction;
  entity_type: string;
  entity_id: number | null;
  details: string | null;
  created_at: string;
}
